pennylane/interfaces/execution.py

```diff
@@ -506,19 +516,54 @@ def cost_fn(params, x):
 
     _grad_on_execution = False
 
-    if gradient_fn == "device":
+    if config.use_device_gradient:
+        # must be new device if this is specified as true
+        _grad_on_execution = config.grad_on_execution
+
+        if config.grad_on_execution:
+
+            def execute_fn(internal_tapes):
+                """A partial function that wraps the execute_and_compute_derivatives method of the device.
+
+                Closure Variables:
+                    device: The device to execute on
+                    config: the ExecutionConfig that specifies how to perform the simulations.
+                """
+                return device.execute_and_compute_derivatives(internal_tapes, config)
+
+            gradient_fn = None
+
+        else:
+
+            def execute_fn(internal_tapes) -> Tuple[ResultBatch, Tuple]:
+                """A wrapper around device.execute that adds an empty tuple instead of derivatives.
+
+                Closure Variables:
+                    device: the device to execute on
+                    config: the ExecutionConfig that specifies how to perform the simulations.
+                """
+                return (device.execute(internal_tapes, config), tuple())
+
+            def gradient_fn(internal_tapes):
+                """A partial function that wraps compute_derivatives method of the device.
+
+                Closure Variables:
+                    device: the device to execute on
+                    config: the ExecutionConfig that specifies how to take the derivative.
+                """
+                return device.compute_derivatives(internal_tapes, config)
+
+    elif gradient_fn == "device":
         # gradient function is a device method
 
         # Expand all tapes as per the device's expand function here.
         # We must do this now, prior to the interface, to ensure that
         # decompositions with parameter processing is tracked by the
         # autodiff frameworks.
-        for i, tape in enumerate(tapes):
-            tapes[i] = expand_fn(tape)
+        tapes = [expand_fn(t) for t in tapes]
 
         if gradient_kwargs.get("method", "") == "adjoint_jacobian":
-            mode = "forward" if grad_on_execution else "backward"
-            tapes = _adjoint_jacobian_expansion(tapes, mode, interface, max_expansion)
+            tapes = _adjoint_jacobian_expansion(tapes, grad_on_execution, interface, max_expansion)

```

Comment:
what's happened here? were we passing `mode` (a string) to a function that was expecting a bool? how did that ever work?